src/libbson/src/bson/bson-json.c

```diff
@@ -362,20 +362,27 @@ _noop (void)
       return;                                                                                                \
    } else                                                                                                    \
       (void) 0
-#define HANDLE_OPTION(_selection_statement, _key, _type, _state)                                 \
-   _selection_statement (len == strlen (_key) && strncmp ((const char *) val, (_key), len) == 0) \
-   {                                                                                             \
-      if (bson->bson_type && bson->bson_type != (_type)) {                                       \
-         _bson_json_read_set_error (reader,                                                      \
-                                    "Invalid key \"%s\".  Looking for values "                   \
-                                    "for type \"%s\", got \"%s\"",                               \
-                                    (_key),                                                      \
-                                    _bson_json_type_name (bson->bson_type),                      \
-                                    _bson_json_type_name (_type));                               \
-         return;                                                                                 \
-      }                                                                                          \
-      bson->bson_type = (_type);                                                                 \
-      bson->bson_state = (_state);                                                               \
+
+#define HANDLE_OPTION_KEY_COMPARE(_key) (len == strlen (_key) && memcmp (key, (_key), len) == 0)
+
+#define HANDLE_OPTION_TYPE_CHECK(_key, _type)                               \
+   if (bson->bson_type && bson->bson_type != (_type)) {                     \
+      _bson_json_read_set_error (reader,                                    \
+                                 "Invalid key \"%s\".  Looking for values " \
+                                 "for type \"%s\", got \"%s\"",             \
+                                 (_key),                                    \
+                                 _bson_json_type_name (bson->bson_type),    \
+                                 _bson_json_type_name (_type));             \
+      return;                                                               \
+   }                                                                        \
+   ((void) 0)
```

Comment:
Should there be an `else` before the cast-to-void?

Reply:
I omitted the `else` because it didn't appear to be necessary. The `if` is scoped with `{}`, so there is (AFAIK) no opportunity for `else` hijacking or other (plausible) mistakes which may produce unexpected behavior at usage sites. A plain `((void)0)` seems sufficient to force a semicolon regardless of the macro definition (`if` or otherwise). If I am incorrect and it is advisable to keep the `else` anyways, a followup PR can patch it in.

Reply:
Since this is so contained, I don't think there's any risk of anything goofy happening. My only thought was about:

```c
if (x)
    HANDLE_OPTION_TYPE_CHECK(...)
else
    lol();
```

But that would just be a hard compile-error anyway.